### DISCUSSION TITLE
installation help (tools.py and reduction.py syntax errors)

### DISCUSSION THREAD

**Discussion by olagarde** [inducer/pycuda#316] [Troubleshooting]
Syntax errors (though they don't stop the install) for the following two files.  I know I've missed the mako prereq, is this an artifact of that or something else I've borked?

- File "build/bdist.linux-x86_64/egg/pycuda/reduction.py", line 459:   reduce_expr=f**_"_**{reduce_expr}"
- File "build/bdist.linux-x86_64/egg/pycuda/tools.py", line 98:   file**_=_**self.logfile

Build/install system:
- python 2.7 (via python-2.7.5-90.el7.x86_64 from CentOS 7 Base repo)
- gcc 11.2.0 (via source build from GitHub stable 11.2.0, defaults, successful 'make check',
        built by gcc 4.8.5 via gcc-4.8.5-44.el7.x86_64 from CentOS 7 Base repo)
- pycuda 2021.1 (from here)
- no multilib, only lib64 on the local system
- local system is a private compute cluster member
        roofs is readonly, install relocated to /opt/soft/cuda/2021.2
        local system can not contact external services (no wget, curl, etc)

Steps taken:
- pycuda downloaded elsewhere, from https://files.pythonhosted.org/packages/5a/56/4682a5118a234d15aa1c8768a528aac4858c7b04d2674e18d586d3dfda04/pycuda-2021.1.tar.gz
- checksums downloaded elsewhere, from https://pypi.org/project/pycuda/#copy-hash-modal-6ab4c3e6-c78b-4d5c-aa8c-e2be3431e4c7
- ... sha256sums matched ...
- tar -xzvpf pycuda-2021.1.tar.gz
- cd pycuda-2021.1
- rm -f siteconf.py
- export PATH=/opt/soft/gcc/11.2.0/bin:$PATH
- export PYTHONPATH=/opt/soft/pycuda/2021.1/lib64/python
- mkdir -p /opt/soft/pycuda/2021.1/lib64/python
- find /opt/soft/pycuda -type d -maxdepth 4 -exec chmod 0755 {} \;
- python ./configure.py --cuda-inc-dir=/usr/include/cuda
- python ./setup.py build
- python ./setup.py install --home=/opt/soft/pycuda/2021.1

Build is successful without significant warnings, just the usual C++{std} compliance, ISO compliance, and template-depricated messages.  Full output attached.  Ignore the mako errors, they happened because I derped the prereqs and the build environment per Customer dictate isn't allowed to talk outside of the clusterdomain.  My question is only about the syntax errors below.
[install.txt](https://github.com/inducer/pycuda/files/7463525/install.txt)

------------
On install everything is fine until bytecompiling reduction.py and tools.py:

        ...
    byte-compiling build/bdist.linux-x86_64/egg/pycuda/reduction.py to reduction.pyc
      File "build/bdist.linux-x86_64/egg/pycuda/reduction.py", line 459
        reduce_expr=f"{reduce_expr}",
                               ^
    SyntaxError: invalid syntax

    byte-compiling build/bdist.linux-x86_64/egg/pycuda/scan.py to scan.pyc
    byte-compiling build/bdist.linux-x86_64/egg/pycuda/tools.py to tools.pyc
      File "build/bdist.linux-x86_64/egg/pycuda/tools.py", line 98
        file=self.logfile,
            ^
    SyntaxError: invalid syntax
       ...

And again when unrolling the egg into the target:

    ...
    Extracting pycuda-2021.1-py2.7-linux-x86_64.egg to /opt/soft/pycuda/2021.1/lib64/python
      File "/opt/soft/pycuda/2021.1/lib64/python/pycuda-2021.1-py2.7-linux-x86_64.egg/pycuda/reduction.py", line 459
        reduce_expr=f"{reduce_expr}",
                               ^
    SyntaxError: invalid syntax

      File "/opt/soft/pycuda/2021.1/lib64/python/pycuda-2021.1-py2.7-linux-x86_64.egg/pycuda/tools.py", line 98
        file=self.logfile,
            ^
    SyntaxError: invalid syntax
    ...

Followed by deps fetch failing because I borked the prereqs, also the build environment per Customer dictate is not allowed to talk outside of the clusterdomain (ignore this in the output, my question is about the syntax error above):

    ...
    Processing dependencies for pycuda==2021.1
    Searching for mako
    Reading https://pypi.python.org/simple/mako/
    Couldn't find index page for 'mako' (maybe misspelled?)
    Scanning index of all packages (this may take a while)
    Reading https://pypi.python.org/simple/
    No local packages or download links found for mako
    error: Could not find suitable distribution for Requirement.parse('mako')

**Comment by inducer** [comment]
Python 2.7 is no longer supported by current PyCUDA. According to the package index, version 2020.1 is the last version to support 2.7: https://pypi.org/project/pycuda/2020.1/

**Comment by olagarde** [comment]
Sorry, lost here ... that url says 2020.1 requires python ~=3.6 not 2.7, 
building 2020.1 or 2021.1 with 3.8.6 thows an error to the effect that 
it requires >3.7, and the wiki says 2021.1 requires 2.4 ... who's on first?

**Comment by olagarde** [comment]
In hindsight I suppose the answer is obvious:  don't depend on the base 
os provisioning for python, it stinks.  Build a properly recent python 
standalone for use with pycuda, pytorch, tensorflow, etc....
